File: src/prod/src/data/txnreplicator/loggingreplicator/ReplicatorBackup.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace Data::LoggingReplicator;
using namespace TxnReplicator;
using namespace Data::Utilities;

LONG64 ReplicatorBackup::InvalidLogSize = -1;
LONG64 ReplicatorBackup::InvalidDataLossNumber = -1;
LONG64 ReplicatorBackup::InvalidConfigurationNumber = -1;

#define NumberOfBytesInKB 1024.0
#define NumberOfBytesInMB 1048576.0

ReplicatorBackup ReplicatorBackup::Invalid()
{
    return ReplicatorBackup();
}

ULONG32 ReplicatorBackup::get_LogCount() const
{
    return logCount_;
}

LONG64 ReplicatorBackup::get_AccumulatedLogSizeInKB() const
{
    return static_cast<LONG64>(ceil(accumulatedLogSize_ / NumberOfBytesInKB));
}

LONG64 ReplicatorBackup::get_AccumulatedLogSizeInMB() const
{
    return static_cast<LONG64>(ceil(accumulatedLogSize_ / NumberOfBytesInMB));
}

FABRIC_EPOCH ReplicatorBackup::get_EpochOfFirstBackedUpLogRecord() const
{
    return epochOfFirstLogicalLogRecord_;
}

FABRIC_SEQUENCE_NUMBER ReplicatorBackup::get_LsnOfFirstLogicalLogRecord() const
{
    return lsnOfFirstLogicalLogRecord_;
}

FABRIC_EPOCH ReplicatorBackup::get_EpochOfHighestBackedUpLogRecord() const
{
    return epochOfHighestBackedUpLogRecord_;
}

FABRIC_SEQUENCE_NUMBER ReplicatorBackup::get_LsnOfHighestBackedUpLogRecord() const
{
    return lsnOfHighestBackedUpLogRecord_;
}

ReplicatorBackup::ReplicatorBackup()
    : logCount_(0)
    , logSize_(InvalidLogSize)
    , accumulatedLogSize_(InvalidLogSize)
    , lsnOfFirstLogicalLogRecord_(FABRIC_INVALID_SEQUENCE_NUMBER)
    , lsnOfHighestBackedUpLogRecord_(FABRIC_INVALID_SEQUENCE_NUMBER)
{
    FABRIC_EPOCH invalidEpoch;
    invalidEpoch.DataLossNumber = InvalidDataLossNumber;;
    invalidEpoch.ConfigurationNumber = InvalidConfigurationNumber;

    epochOfFirstLogicalLogRecord_ = invalidEpoch;
    epochOfHighestBackedUpLogRecord_ = invalidEpoch;
}

ReplicatorBackup::ReplicatorBackup(
    __in ULONG32 logCount,
    __in LONG64 logSize,
    __in LONG64 accumulatedLogSize,
    __in FABRIC_EPOCH epochOfFirstLogicalLogRecord,
    __in FABRIC_SEQUENCE_NUMBER lsnOfFirstLogicalLogRecord,
    __in FABRIC_EPOCH epochOfHighestBackedUpLogRecord,
    __in FABRIC_SEQUENCE_NUMBER lsnOfHighestBackedUpLogRecord)
    : logCount_(logCount)
    , logSize_(logSize)
    , accumulatedLogSize_(accumulatedLogSize)
    , epochOfFirstLogicalLogRecord_(epochOfFirstLogicalLogRecord)
    , lsnOfFirstLogicalLogRecord_(lsnOfFirstLogicalLogRecord)
    , epochOfHighestBackedUpLogRecord_(epochOfHighestBackedUpLogRecord)
    , lsnOfHighestBackedUpLogRecord_(lsnOfHighestBackedUpLogRecord)
{
}

ReplicatorBackup & ReplicatorBackup::operator=(
    __in ReplicatorBackup const & other)
{
    if (&other == this)
    {
        return *this;
    }

    logCount_ = other.logCount_;
    logSize_ = other.logSize_;
    accumulatedLogSize_ = other.accumulatedLogSize_;

    epochOfFirstLogicalLogRecord_ = other.epochOfFirstLogicalLogRecord_;
    lsnOfFirstLogicalLogRecord_ = other.lsnOfFirstLogicalLogRecord_;
    epochOfHighestBackedUpLogRecord_ = other.epochOfHighestBackedUpLogRecord_;
    lsnOfHighestBackedUpLogRecord_ = other.lsnOfHighestBackedUpLogRecord_;

    return *this;
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/ReplicatorBackup.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Fabric;

    internal class ReplicatorBackup
    {
        private readonly long accumulatedLogSize;

        private readonly Epoch epochOfFirstLogicalLogRecord;

        private readonly Epoch epochOfHighestBackedUpLogRecord;

        private readonly uint logCount;

        private readonly long logSize;

        private readonly LogicalSequenceNumber lsnOfFirstLogicalLogRecord;

        private readonly LogicalSequenceNumber lsnOfHighestBackedUpLogRecord;

        public ReplicatorBackup(
            Epoch epochOfFirstLogicalLogRecord,
            LogicalSequenceNumber lsnOfFirstLogicalLogRecord,
            Epoch epochOfHighestBackedUpLogRecord,
            LogicalSequenceNumber lsnOfHighestBackedUpLogRecord,
            uint logCount,
            long logSize,
            long accumulatedLogSize)
        {
            this.epochOfFirstLogicalLogRecord = epochOfFirstLogicalLogRecord;
            this.lsnOfFirstLogicalLogRecord = lsnOfFirstLogicalLogRecord;

            this.epochOfHighestBackedUpLogRecord = epochOfHighestBackedUpLogRecord;
            this.lsnOfHighestBackedUpLogRecord = lsnOfHighestBackedUpLogRecord;

            this.logCount = logCount;
            this.logSize = logSize;
            this.accumulatedLogSize = accumulatedLogSize;

            Utility.Assert(accumulatedLogSize >= this.logSize, "accumulated log size is larger.");
        }

        public long AccumulatedLogSize
        {
            get { return this.accumulatedLogSize; }
        }

        public uint AccumulatedLogSizeInKB
        {
            get { return (uint) Math.Ceiling(this.accumulatedLogSize/(double) 1024); }
        }

        public long AccumulatedLogSizeInMB
        {
            get { return (long) Math.Ceiling(this.accumulatedLogSize/((double) 1024*(double) 1024)); }
        }

        public Epoch EpochOfFirstLogicalLogRecord
        {
            get { return this.epochOfFirstLogicalLogRecord; }
        }

        public Epoch EpochOfHighestBackedUpLogRecord
        {
            get { return this.epochOfHighestBackedUpLogRecord; }
        }

        public uint LogCount
        {
            get { return this.logCount; }
        }

        public long LogSize
        {
            get { return this.logSize; }
        }

        public LogicalSequenceNumber LsnOfFirstLogicalLogRecord
        {
            get { return this.lsnOfFirstLogicalLogRecord; }
        }

        public LogicalSequenceNumber LsnOfHighestBackedUpLogRecord
        {
            get { return this.lsnOfHighestBackedUpLogRecord; }
        }
    }
}